MVC step2 [띠용]

jye326 opened the pull request:
> 미션 PR 타이틀 양식이 있었나 해서 일단 그냥 써봤습니다. 문제가 되면 변경하겠습니다.

- 이번 목표는 '쉽게 읽히는 코드'입니다. 가독성 위주의 피드백도 환영합니다!

- 동작을 이해하시는데 도움이 되도록 제가 이해한 개념들도 일부 주석으로 첨부했어요!

- step1에서 칼리의 날카로운 질문들 덕분에 많은 부분에서 더 깊게 고민할 수 있었어요! 이번에도 열정적인 리뷰 부탁 드릴게요~!

jye326 committed:
test: 리플렉션 테스트 수정 및 누락된 메소드 구현

- 클래스 네이밍 테스트 수정
- 필드 및 메소드 추출 로직 구현
- 생성자 관련 테스트 케이스 수정
- Junit3/4TestRunner 실행 로직 구현
- 필요한 테스트 라이브러리 추가

jye326 committed:
feat: 추가 Servlet 구현 및 CharacterEncodingFilter 수정

- CharacterEncodingFilter: response content type 설정 추가
- FormHandlingServlet: GET/POST 요청 처리 로직 구현
- HttpMethodServlet: HTTP 메서드 처리 서비스 구현
- BasicServlet: 기본 Servlet 동작 확인용 구현

jye326 committed:
fix: enable filter chain execution in CharacterEncodingFilter

jye326 committed:
build: JUnit 의존성 구성 변경 및 platform BOM 추가

jye326 committed:
feat: HandlerExecution 클래스에 동적 메서드 호출 추가

- 동적 메서드 호출을 위한 handler와 method 필드 추가
- 메서드 접근 허용을 위해 setAccessible(true) 호출
- HttpServletRequest와 HttpServletResponse를 인자로 받는 메서드 호출 로직 구현

jye326 committed:
feat: AnnotationHandlerMapping 클래스에 어노테이션 기반 핸들러 매핑 추가

- @Controller와 @RequestMapping 어노테이션 기반으로 핸들러 자동 스캔 구현
- HandlerKey와 HandlerExecution을 활용한 핸들러 등록 로직 추가
- HTTP 요청 메서드와 URI 매핑을 통한 핸들러 검색 기능 구현
- URI 정규화 메서드 추가

jye326 committed:
feat: JspView 클래스 구현 및 DispatcherServlet에 적용

- JspView에서 JSP 뷰 렌더링 및 리다이렉트 처리 로직 추가
- DispatcherServlet에 JspView를 이용한 뷰 렌더링 로직 수정
- 불필요한 move 메서드 제거 및 관련 뷰 처리 로직 간소화

jye326 committed:
refactor: AnnotationHandlerMapping 메서드 분리 및 코드 가독성 개선

- initialize 메서드 내부 로직 handlerMapping 메서드로 분리
- stream 사용으로 메서드 매핑 로직 간결화
- 로깅 추가 및 코드 구조 정리

jye326 committed:
refactor: JspView 리다이렉트 처리 로직 분리 및 ViewResolver 도입

- JspView 내부 리다이렉트 처리 로직 ViewResolver로 이동
- DispatcherServlet에서 ViewResolver를 활용한 리다이렉트 처리 추가
- 코드 구조 개선 및 가독성 향상
- 주석 수정 및 필요 없는 import 정리

jye326 committed:
refactor: Controller 인터페이스 리팩토링 및 반환 타입 수정

- Controller 인터페이스 execute 메서드 반환 타입 변경 (String → ModelAndView)
- View 반환 로직에서 JspView 객체로 전환
- 가독성 향상을 위한 코드 최적화 및 로깅 추가

jye326 committed:
refactor: ControllerScanner 도입 및 AnnotationHandlerMapping 리팩토링

- @Controller 스캔 및 인스턴스 매핑 역할 ControllerScanner로 분리
- initialize 메서드 리팩토링 및 매핑 로직 간소화
- ReflectionUtils 활용으로 메서드 검색 및 핸들러 매핑 코드 최적화
- 코드 구조 개선 및 로깅 수정

jye326 committed:
refactor: HandlerAdapter 도입 및 DispatcherServlet 리팩토링

- 핸들러 처리를 위한 HandlerAdapter 클래스 추가
- DispatcherServlet 일부 로직 HandlerAdapter로 위임
- AnnotationHandlerMapping 초기화 로직 및 핸들러 등록 추가
- 예외 상황을 처리하기 위한 NoHandlerFoundException 클래스 추가
- JspView 리다이렉트 처리 조건 추가 및 코드 정리

jye326 committed:
Merge branch 'step2' of https://github.com/jye326/java-mvc into step2

# Conflicts:
#	app/src/main/java/com/techcourse/DispatcherServlet.java
#	mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java
#	mvc/src/main/java/com/interface21/webmvc/servlet/view/JspView.java

jye326 committed:
refactor: ModelAndView 객체 생성 로직 간소화

- ModelAndView.ofJspView 정적 메서드 추가
- 기존 new JspView 인스턴스 생성 코드 변경 및 중복 제거
- 불필요한 import문 정리

jye326 committed:
refactor: ControllerScanner 클래스 주석 개선

- 인스턴스 매핑 관련 주석 내용 명확화

jye326 committed:
refactor: JspView 및 AnnotationHandlerMapping 주석 개선

- JspView 클래스의 주석 줄바꿈 수정
- AnnotationHandlerMapping 클래스에 initialize 메서드 관련 주석 추가 및 리플렉션 기반 매핑 로직 설명 보완

jye326 committed:
refactor: HandlerMapping 인터페이스에 initialize 메서드 추가 및 ManualHandlerMapping 구현 수정

- HandlerMapping 인터페이스에 initialize 메서드 정의
- ManualHandlerMapping 클래스에 initialize 메서드 구현 및 Override 추가

jye326 committed:
refactor: DispatcherServlet 및 HandlerAdapter 코드 간소화

- handlerMappings 초기화 방식 변경 및 반복 초기화 제거
- HandlerAdapter 메서드 시그니처에서 HandlerMapping 의존성 제거
- 적절한 핸들러를 찾지 못한 경우 404 에러 응답 처리 추가
- service 메서드 내 핸들러 매핑 탐색 로직 명확화 및 주석 추가

jye326 committed:
refactor: DispatcherServlet 핸들러 검색 로직 간소화

- 불필요한 handler null 초기화 코드 제거
- 핸들러 매핑에서 적절한 핸들러 발견 시 즉시 반환하도록 수정
- 404 응답 처리 로직 간결화

jye326 committed:
refactor: UriUtils 유틸 클래스 추가 및 URI 처리 로직 개선

- ManualHandlerMapping에서 URI 처리 로직 UriUtils로 이동
- ResourceFilter의 URI 처리 코드 개선 및 UriUtils 활용
- URI 처리 전용 유틸리티 메서드 getResourcePath 추가

jye326 committed:
refactor: ControllerScanner의 getControllers 메서드 반환값 수정

- 반환 시 새로운 HashMap 인스턴스를 생성하여 불변성 확보

jye326 committed:
refactor: HandlerAdapter의 render 메서드 호출 시 Model 전달 방식 수정

- Map.of() 대신 ModelAndView의 getModel로 모델 객체 전달
- 불필요한 Map import 제거

jye326 committed:
refactor: DispatcherServlet 핸들러 매핑 초기화 순서 수정

- ManualHandlerMapping 초기화 순서를 AnnotationHandlerMapping 이후로 조정

jye326 committed:
refactor: AnnotationHandlerMapping 패키지 경로 수정

- 파일 경로 구분자를 "/"에서 "."으로 변경하여 표준 패키지 표현 방식 사용

jye326 commented:
@coderabbitai 최대한 간결하게 코드 리뷰 해 줘. 모든 리뷰는 한 문장으로 끝내.

2Jin1031 commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/HandlerAdapter.java:
```diff
@@ -0,0 +1,21 @@
+
+public class HandlerAdapter {
+    public static void handle(HttpServletRequest request, HttpServletResponse response,
+                               Object handler) throws Exception {
+        ModelAndView mav;
+        if (handler instanceof Controller) {
+            mav = ((Controller)handler).execute(request, response);
+        } else if (handler instanceof HandlerExecution) {
+            mav = ((HandlerExecution)handler).handle(request, response);
+        } else {
+            throw new NoHandlerFoundException("No Handler Found!");
+        }
+        mav.getView().render(mav.getModel(), request, response);
+    }
```
요구사항에 따라 handlerAdapter를 인터페이스화하는 방향도 괜찮을 것 같아요
핸들러에 적합하는 지 확인인 후, 해당 Adapter가 handle를 수행하는 구조는 어떨까요~?

 handlerAdater의 실제 구현 형태가 궁금하다면
이 블로그를 참고해봐도 좋을 것 같아요~
[HandlerMapping과 HandlerAdapter는 왜 나뉘었나요?](https://stonehee99.tistory.com/24)

jye326 replied:
1. 요구사항 대로 `handlerAdapter`를 인터페이스로 만들었었는데 만들고 보니 '왜 인터페이스여야 하지?' 싶어 일반 클래스로 구현했습니다.
2. 만약 인터페이스로 두려면 핸들러의 종류를 사전에 파악하여 적절한 어댑터 구현체를 선택해야 할 텐데, 이런 분기 로직이 디스패처 서블릿에 있는 것이 부자연스럽다고 생각했습니다. 차라리 예제 코드처럼 단일 `handlerAdapter` 내부에서 핸들러 종류에 따라 분기 처리를 하게 되면 디스패처 서블릿의 책임을 덜 수 있다고 생각했습니다.
3. 그런데 왜 SpringMVC에서는 어댑터를 인터페이스로 두었을까요?
4. 인터페이스가 아닌 단일 클래스 `handlerAdapter` 에서 핸들러의 종류가 100가지 생겼을 때의 내부 분기문을 상상하니 인터페이스의 필요성을 공감할 수 있었습니다.
5. 그러나 디스패처 서블릿에서 분기문으로 적절한 구현체를 탐색하면 기존과 동일한 복잡함이 생길 것입니다.
6. 적절한 구현체를 탐색 방법을 고민하다 `handlerMapping` 구현체를 선택하는 것과 동일한 방식으로 `handlerAdapter` 구현체를 선택할 수 있다는 것을 깨달았습니다.
7. 이제 새로운 HandlerAdapter 구현체가 추가될 때 디스패처 `handlerAdapters`에 구현체를 추가하기만 하면 되겠네요! 하하~

2Jin1031 replied:
역시 ㄷㄷ.. 탁월한 띠용이네요

2Jin1031 commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/ControllerScanner.java:
```diff
@@ -0,0 +1,37 @@
+        Map<Class<?>, Object> controllers = new HashMap<>();
+        classSet.forEach(clazz -> {
+            try {
+                Object o = clazz.getDeclaredConstructor().newInstance();
```
`controllers`라는 변수명은 보통 여러 컨트롤러 객체가 단순히 모여 있는 List 같은 느낌을 줄 수 있을 것 같아요 
실제로 `Map<Class<?>, Object>` 구조라 의미가 조금 다르다고 생각했어요 그래서 이 구조가 가진 의도를 더 잘 들어낼 수 있는 이름을 쓰면 코드가 읽히는 맥락이 한결 명확해질 것 같더라구요
특히 자료형을 직접 노출하지 않고도 의미를 담을 수 있다면 훨씬 자연스러울 것 같아요~

생성된 인스턴스도 마찬가지로요!

jye326 replied:
고민에 고민을 거듭한 끝에 `controllerRegistry`로 변경하였습니다! 참 잘 지었지요? 하하

2Jin1031 replied:
이또한 딱띠네요

2Jin1031 commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,40 +23,61 @@ public AnnotationHandlerMapping(final Object... basePackage) {
-        }
+        ControllerScanner controllerScanner = new ControllerScanner(basePackage);
+        // @Controller 클래스 - 인스턴스 맵
+        Map<Class<?>, Object> controllers = controllerScanner.getControllers();
+        controllers.keySet().forEach(clazz -> {
+            Set<Method> allMethods = ReflectionUtils.getAllMethods(clazz,
+                    ReflectionUtils.withAnnotation(RequestMapping.class));
+            mappingAllHandlerMethods(clazz, allMethods, controllers);
+        });
```
`keySet` 으로 반복문을 돌리고 있어 `Map`의 `key`와 `value`가 동시에 필요한 상황인 `mappingAllHandlerMethods`에 `map`을 함께 넘겨줘야 하는 상황인 것 같아요!

`key`와 `value`를 한 번에 다룰 수 있는 `entrySet()`을 이용해봐도 좋을 것 같아요~

jye326 replied:
역시 칼리...!👏 천재적인 방법이네요! 반영했습니다!

2Jin1031 reviewed (changes requested):
step2 고생했어요~ 띠용!
요구사항을 잘 만족시켜주셔서 따로 코멘트 드릴게 별로 없더라구요

간단한 코멘트들만 조금 남겨놨으니 편하게 확인하시고 다시 요청해주시면
approve & merge 진행할게요 :)

jye326 committed:
refactor: support 메서드 추가 및 DispatcherServlet 처리 로직 개선

- HandlerMapping 인터페이스에 support 메서드 추가
- AnnotationHandlerMapping 및 ManualHandlerMapping에 support 메서드 구현
- DispatcherServlet에서 핸들러 처리 조건에 support 메서드 적용
- 코드 가독성 향상을 위한 로직 단순화 및 리팩토링

jye326 committed:
refactor: HandlerAdapter 인터페이스 분리 및 DispatcherServlet 리팩토링

- HandlerAdapter 인터페이스로 리팩토링 및 클래스 분리
- ControllerAdapter, HandlerExecutionAdapter 구현체 추가
- DispatcherServlet에서 handler 처리 로직을 handlerAdapters로 위임
- 코드 구조 개선 및 가독성 향상

jye326 committed:
refactor: 예외 클래스 리팩토링 및 패키지 이동

- NoHandlerFoundException 등 예외 클래스 패키지를 exception으로 이동
- NoHandlerAdapterFoundException 및 NoHandlerMappingFoundException 클래스 추가
- 예외 처리 구분을 위한 클래스 분리 및 가독성 향상

jye326 committed:
refactor: ControllerScanner 필드명 변경 및 AnnotationHandlerMapping 코드 가독성 개선

- controllers 필드명을 controllerRegistry로 변경
- 관련 메서드 및 변수명 일괄 수정
- AnnotationHandlerMapping의 initialize 메서드 내 코드 간결화
- 불필요한 코드 제거 및 가독성 향상

jye326 committed:
refactor: AnnotationHandlerMapping 메서드 시그니처 변경 및 코드 개선

- mappingAllHandlerMethods 매개변수 수정 (class → entry 활용)
- 메서드 호출 로직 간결화 및 코드 가독성 향상
- 불필요한 변수 제거 및 개선된 로직 적용

jye326 reviewed (commented):
모든 코멘트 반영했습니다! 하하~

jye326 committed:
refactor: DispatcherServlet 및 불필요한 HandlerAdapterrrr 제거

- HandlerMapping 및 HandlerAdapter 탐색 로직을 메서드로 분리하여 가독성 향상
- 사용하지 않는 HandlerAdapterrrr 클래스 삭제
- 404 처리 시 응답 코드 전송 및 예외 메시지 명확화

jye326 committed:
refactor: HandlerAdapter의 handle 메서드 반환값 수정 및 DispatcherServlet 로직 개선

- HandlerAdapter의 handle 메서드 반환 타입을 ModelAndView로 변경
- DispatcherServlet에서 ModelAndView 반환 처리 및 View 렌더링 로직 추가
- ModelAndView의 모델 객체 활용을 위한 Map 타입 import 추가

jye326 committed:
refactor: AnnotationHandlerMapping의 URI 처리 로직 개선 및 UriUtils 추가

- URI 처리 코드를 UriUtils.getResourcePath 메서드로 분리
- AnnotationHandlerMapping에서 UriUtils 활용하여 요청 URI 처리

jye326 commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,40 +23,65 @@ public AnnotationHandlerMapping(final Object... basePackage) {
+        }
+    }
+
+    @Override
+    public Object getHandler(HttpServletRequest request) {
         String uri = normalize(request.getRequestURI());
         return handlerExecutions.get(new HandlerKey(uri, RequestMethod.valueOf(request.getMethod())));
     }
```
오케이~

jye326 committed:
refactor: DispatcherServlet의 예외 처리 및 404 응답 로직 개선

- NoHandlerMappingFoundException 발생 시 404 응답 코드 전송 처리 추가
- service 메서드의 throws 선언에 IOException 추가
- 기존 404 응답 처리 중복 코드 제거

jye326 commented on mvc/src/main/java/com/interface21/webmvc/util/UriUtils.java:
```diff
@@ -0,0 +1,9 @@
+import jakarta.servlet.http.HttpServletRequest;
+
+public class UriUtils {
+    public static String getResourcePath(HttpServletRequest req) {
+        return req.getRequestURI().substring(req.getContextPath().length());
+    }
```
굳

jye326 committed:
refactor: AnnotationHandlerMapping의 URI 정규화 로직 수정 및 UriUtils 개선

- URI 정규화를 UriUtils.normalize 메서드로 통합
- AnnotationHandlerMapping에서 요청 URI 처리 시 정규화 적용
- UriUtils에 URI 정규화 메서드 추가 및 getResourcePath 활용 방식 개선

2Jin1031 reviewed (approved):
고생하셨습니다 띠용
모든 코멘트를 반영해주셨네요 ㅎ

충분히 요구사항을 만족해주신 것 같아 approve & merge 진행할게요
마지막 step3도 화이팅 입니다~!!!

+) 아 추가로 commit content가 참 알차더라구요 저에게만 비법 전수해주시면 감사하겠습니다
